Sync v2-7-stable with v2-7-test to release 2.7.0

ephraimbuddy opened the pull request:
Time for `2.7.0rc2`

potiuk committed:
Add update-constraints command that allows to modify released ones (#33144)

* Add update-constraints command that allows to modify released ones

This command can be used to semi-automatically update constraints
in case we need to update historical version of them, because for
some reason (like setuptools or pip or cython changes) we need
to update individual constraints in historically released constraints.

---------

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>
(cherry picked from commit dcfca57eb53840d396ceec25e73c09204bd65d35)

pankajkoti committed:
Enable test connection in UI if explicitly set to `Enabled` (#33342)

Following up PR https://github.com/apache/airflow/pull/32052/ the test connection is
disabled in UI, API and CLI. The API and CLI strictly check for the
config value to be set as `Enabled` for the functionality to be enabled,
whereas the UI just checks that is it not set to `Disabled`. As a result
setting values to the config param other than `Disabled`, enables the
button in the UI. Even though the button gets enabled, the API forbids
it as there is a strict check in the API that the value is set to
`Enabled` and only then allows, however, it makes sense to also strictly
check in the UI that value is set to `Enabled`.

(cherry picked from commit 50765eb0883652c16b40d69d8a1ac78096646610)

eladkal committed:
Rename `ExternalBranchPythonOperator` to `BranchExternalPythonOperator` (#33360)

(cherry picked from commit 9a1624b93d97b3a62cee1efbe1c520699e1323ee)

potiuk committed:
Limit aiobotocore for eager upgrade to prevent pip backtracking (#33362)

The new aiobotocore release 2.6.0 caused `pip` to backtract when
finding `good` set of dependencies. This limitation should stop
the backtracking until other dependncies causing the conflicts
will be updated.

(cherry picked from commit c6c8c1afa5e0a04e8921ce9cb4979c3120b2d5be)

dstandish committed:
Fix clearing behavior for downstream work task with non-collinear setup task (#33358)

* Fix clearing behavior for downstream work task with non-collinear setup

With this kind of dag if you clear w1 downstream then you also clear w2:

```python
    s1 >> w1 >> [w2, t1]
    s1 >> t1
    s2 >> t2
    s2 >> w2 >> t2
```

We need to make sure that the setup for w2 also gets cleared.  But, to avoid the need to recurse to arbitrary depth for setups of setups, let's just say that a setup cannot have a setup.  A setup can *come after* another setup, but it won't *be* a setup for the setup (and what's at stake is just the clearing behavior).

* fixup

* teardowns can't have setup / teardown either

(cherry picked from commit 45713446f37ee4b1ee972ab8b5aa1ac0b2482197)

jscheffl committed:
Make skip of trigger form in UI if no params are defined configurable (#33351)

* Make skip of trigger form in UI if no params are defined configurable

* Review feedback, remove negating bool

* Review feedback, remove negating bool

* Add newsfragment

(cherry picked from commit c0362923fd8250328eab6e60f0cf7e855bfd352e)

millsks committed:
Modify pathspec version restriction (#33349)

Allow pathspec >=0.9.0 to meet requirements of other packages that depend on newer versions of pathspec.  For example, DVC requires pathspec >=0.10.3.

(cherry picked from commit 699b3f4d58f52d9dde1bc063d6f84b1982ccd1db)

jedcunningham committed:
Clarify UI user types in security model (#33021)

* Clarify op users in security model

* Switch the order around

(cherry picked from commit f1fad44650e247b03f79e7db8d70d334e74468a7)

jedcunningham committed:
Fix links in security docs (#33329)

(cherry picked from commit 0cb256411a02516dc9eca88b570abfb8c8a3c35b)

potiuk committed:
Remove user sessions when resetting password (#33347)

* Remove user sessions when resetting password

When user's password is reset, we also remove all DB sessions
for that user - for database session backend.

In case we are using securecookie mechanism, resetting password does
not invalidate old sessions, so instead we are displaying warning to
the user performing the reset that in order to clear existing
sessions of the user, the secure_key needs to be changed and it
will invalidate all sessions for all users.

Protection has been added in case the number of sessions in the DB
is too big to effectively scan and remove sessions for the user. In
such case we print warning for the user that sessions have not
been reset, and we suggest to improve the way their deployment
mechanisms create too many sessions - by either changing the way
how automation of the API calls is done and/or by purging the
sessions regularly by "airflow db clean".

* Update airflow/auth/managers/fab/security_manager/override.py

Co-authored-by: Hussein Awala <hussein@awala.fr>

---------

Co-authored-by: Hussein Awala <hussein@awala.fr>
(cherry picked from commit 2caa186935151683076b74357daad83d2538a3f6)

potiuk committed:
Improve handling of CI image timeout when backtracking (#33364)

Even the latest pip can enter into a long loop of backtracking
when trying to find latest "good" set of dependencies with
eager upgrade. This happened on August 10th 2023 with aiobotocore
causing backtracking.

This PR adds a complete set of tools and instructions that can
help in such cases and figure out which newly released dependency
causes backtracking.

The toolset consists of:

* adding timeout on the image build, so that it can fail before
  the job timeout and provide useful instructions what to do

* adding `ci find-backtracking-candidates` that allows to identify
  the packages released after the last successful constraint update
  that could be the reason for backtracking

* running the `find-backtracking-candidates` command in the CI
  when timeout occurs - this will help to see the candidates as
  early as possible - at the first build that will fail with
  timeout. This should help with narrowing down the root cause
  much faster

* adding detailed explanation why we have the problem and how to
  deal with it step-by-step, including example based on the
  August 2023 backtracking issue with aiobotocore

* finally removing `--empty-image` switch and pushing empty images
  in CI. This was an attempt to speed up waiting for image in case
  the image failed, but what it did, it has hidden the failures
  of the images when they failed. It does not really add value
  any more, since "image waiting" is now always done using small
  public runners, waiting till timeout for those is not a big issue.

(cherry picked from commit 1cf960d5d5f9179b1167d9e660142612781ddeda)

potiuk committed:
Restore `--builder` command for cache building (#33387)

The change #33364 removed accidentally --builder command that has been
used in CI to build image cache.

This PR restores it and also moves it to the "cache" option group where
it belongs.

(cherry picked from commit 2ab0d2d683c548aca87d6ea777cb26b513a6bc8d)

ephraimbuddy committed:
fixup! Update RELEASE_NOTES.rst

potiuk committed:
Add better instructions to constraint files (#33367)

Following up after discussion in #33241 this PR adds more
straightforward instructions on how constraints should (or should
not) be used.

Having the instructions in the constraint files increases the
chance that users will use the constraints as intended.

(cherry picked from commit b8cdd284dd6e7197c8045cbce614b83b235d23a9)

hussein-awala reviewed (approved):
Looks good, all the fixes we discussed recently are included!

potiuk committed:
Fix eager upgrade failures for canary builds (#33395)

The canary builds continued failing after fixing aibotocore
in #33364 when aiobotocore dependency limit dependency limit. This was
because the main build is now using "empty" eager upgrade requirements
which override the embedded aiobotocore and continue trigger
backtracking.

This is not easily visible on CI because the backtracking
error is under a folded CI image build group in CI.

This PR fixes both:

* only overrides eager requirements if they are not empty
* make sure that eager upgrade output is in the unfolded output
  of CI job

(cherry picked from commit 3857d3399c2e5f4c3e0a838b7a76296c4aa19b3e)

ephraimbuddy committed:
Bring back the decision buttons on DAG trigger (#33394)

When show_trigger_form_if_no_params is set to True, the decision
buttons to trigger with or without config should be shown just like
we had it in 2.6.3

(cherry picked from commit 8cc68e244b511200f5b4ec29e3ae87e211493c28)

potiuk reviewed (approved):
LGTM. There are a few "dev" related small fixes that I merged yesterday (that were about to complete the "eager upgrade" problem resolution  - that will need to be likely cherry-picked. But other than that all good.

pierrejeambrun reviewed (approved):
Nice

jedcunningham commented on RELEASE_NOTES.rst:
```diff
@@ -37,6 +37,13 @@ Old Graph View is removed (#32958)
 The old Graph View is removed. The new Graph View is the default view now.
 
+The trigger UI form is skipped in web UI with 2.7.0 if no parameters are defined in a DAG (#33351)
+""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
```
```suggestion
The trigger UI form is skipped in web UI if no parameters are defined in a DAG (#33351)
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
```

nit

jedcunningham commented on RELEASE_NOTES.rst:
```diff
@@ -332,6 +340,8 @@ Misc/Internal
 
 Docs only changes
 """""""""""""""""
+- Fix links in security docs (#33329)
```
```suggestion
```

Don't need this one in the release notes.

eladkal reviewed (approved):
We need also to wait for a fix for labels issue

https://apache-airflow.slack.com/archives/CCPRP7943/p1692086197633229

bbovenzi committed:
Improve graph nesting logic (#33421)

(cherry picked from commit 6547b8182f056b6d29445dec50cce3920d36f762)
